documents/coredocument.go

```diff
@@ -587,6 +590,36 @@ func (cd *CoreDocument) Timestamp() (time.Time, error) {
 	return utils.FromTimestamp(cd.Document.Timestamp)
 }
 
+// AddAttribute adds a custom attribute to the model with the given value. If an attribute with the given name already exists, it's updated.
+func (cd *CoreDocument) AddAttribute(name string, attributeType AllowedAttributeType, value string) error {
+	// TODO convert value from string to correct type
+	// For now its only string that is supported
+	nAttr, err := newAttribute(name, attributeType, value)
+	if err != nil {
+		return err
+	}
+	if cd.Attributes == nil {
+		cd.Attributes = make(map[string]*attribute)
+	}
+	cd.Attributes[name] = nAttr
+	return nil
+}
+
+// GetAttribute gets the attribute with the given name from the model together with its type, it returns a non-nil error if the attribute doesn't exist or can't be retrieved.
+func (cd *CoreDocument) GetAttribute(name string) (hashedKey []byte, attrType string, value interface{}, valueStr string, err error) {
+	if attr, ok := cd.Attributes[name]; ok {
+		// TODO convert value to its string repr
+		return attr.hashedKey, string(attr.attrType), attr.value, "", nil
+	}
+	return hashedKey, attrType, value, valueStr, errors.NewTypedError(ErrCDAttribute, errors.New("attribute does not exist"))
+}
+
+// DeleteAttribute deletes a custom attribute from the model
+func (cd *CoreDocument) DeleteAttribute(name string) error {
```

Comment:
Any mutations to CoreDocument should return the new model. This will ensure that we dont mutate the current one